### PR TITLE
Reimplement signal handler to avoid async-unsafe functions

### PR DESCRIPTION
**Pull request by zzxuanyuan**
@bbockelm @pcanal This patch copies the code of signal handling in CMSSW. It avoids async-unsafe functions in signal handler functions. 

For reference, see the link https://github.com/bbockelm/cmssw/blob/stacktrace_handler_revisit/FWCore/Services/src/InitRootHandlers.cc

I tried this patch with some simple multi-thread test cases and it worked fine. Is there any complicated test cases I can run? I think this patch is not very ready to merge, but it achieved basic functions. Any criticisms are welcome.

### CODE REVIEW

File: core/unix/inc/TUnixSystem.h [hunk 1]
```diff
@@ -31,11 +31,22 @@
 #include "TTimer.h"
 #endif
 
+#include <thread>
+#include <memory>
+
+namespace std {
+  class thread;
```
Comment (bbockelm) [inline core/unix/inc/TUnixSystem.h]: Don't need a forward dec'l if you already include <thread>

File: core/unix/src/TUnixSystem.cxx [hunk 2]
```diff
@@ -568,12 +568,68 @@ TUnixSystem::~TUnixSystem()
...
+          if (errno == EINTR) {continue;}
+          else {return -errno;}
+        }
+        count -= written;
+        buffer += written;
+      }
+      return 0;
+    }
+
+    static int full_read(int fd, char *inbuf, size_t len)
```
Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: Add in an optional timeout to this function so the thread requesting the stack trace is less likely to timeout.

File: core/unix/src/TUnixSystem.cxx [hunk 3]
```diff
@@ -2154,10 +2210,8 @@ void TUnixSystem::StackTrace()
       }
       gdbscript += " ";
    }
-
```
Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: Don't do unnecessary whitespace changes.

File: core/unix/src/TUnixSystem.cxx [hunk 4]
```diff
@@ -5148,3 +5222,154 @@ int TUnixSystem::GetProcInfo(ProcInfo_t *info) const
...
+      while(true)
+      {
+        int result = full_read(fromParent, buf, 1);
+        if (result < 0)
+        {
+          close(toParent);
+          full_cerr_write("\n\nTraceback helper thread failed to read from parent: ");
+          full_cerr_write(strerror(-result));
+          full_cerr_write("\n");
+          ::abort();
```
Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: Reset signal handler here to SIG_DFL.  Otherwise, ::abort() will deadlock.

File: core/unix/src/TUnixSystem.cxx [hunk 5]
```diff
@@ -5148,3 +5222,154 @@ int TUnixSystem::GetProcInfo(ProcInfo_t *info) const
...
+        {
+          break;
+        }
+        else
+        {
+          close(toParent);
+          full_cerr_write("\n\nTraceback helper thread got unknown command from parent: ");
+          full_cerr_write(buf);
+          full_cerr_write("\n");
+          ::abort();
```
Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: Same as above - reset signal handler.

File: core/unix/src/TUnixSystem.cxx [hunk 6]
```diff
@@ -5148,3 +5222,154 @@ int TUnixSystem::GetProcInfo(ProcInfo_t *info) const
...
+   return pstackArgv_;
+}
+
+void TUnixSystem::cachePidInfo()
+{
+#ifdef ROOTETCDIR
+      if(sprintf(pidString_, "%s/gdb-backtrace.sh", ROOTETCDIR) >= pidStringLength_)
+      {
+        full_cerr_write("Unable to pre-allocate executable information");
+	exit(1);
```
Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: Do not call exit() from a library - instead, make it so this function can fail as necessary.

File: core/unix/src/TUnixSystem.cxx [hunk 7]
```diff
@@ -5148,3 +5222,154 @@ int TUnixSystem::GetProcInfo(ProcInfo_t *info) const
...
+}
+
+char *const *TUnixSystem::getPstackArgv() {
+   return pstackArgv_;
+}
+
+void TUnixSystem::cachePidInfo()
+{
+#ifdef ROOTETCDIR
+      if(sprintf(pidString_, "%s/gdb-backtrace.sh", ROOTETCDIR) >= pidStringLength_)
```
Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: I think you mean to use snprintf here?

Regardless, I think it may be better to switch to std::string and std::stringstream to form this.

File: core/unix/src/TUnixSystem.cxx [hunk 8]
```diff
@@ -5148,3 +5222,154 @@ int TUnixSystem::GetProcInfo(ProcInfo_t *info) const
...
+      {
+        full_cerr_write("Unable to pre-allocate executable information");
+	exit(1);
+      }
+#else
+      if(sprintf(pidString_, "%s/etc/gdb-backtrace.sh", Getenv("ROOTSYS")) >= pidStringLength_)
+      {
+        full_cerr_write("Unable to pre-allocate executable information");
+	exit(1);
```
Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: Use consistent indentation.

File: core/unix/src/TUnixSystem.cxx [hunk 9]
```diff
@@ -568,12 +568,68 @@ TUnixSystem::~TUnixSystem()
...
+    }
+
+    static int full_cerr_write(const char *text)
+    {
+      return full_write(2, text);
+    }
+
+}
+
+static char pshellExec[8] = "/bin/sh";
```
Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: Probably less error-prone to do pshellExec[] instead of pshellExec[8]; let the compiler do the work for you.

File: core/unix/inc/TUnixSystem.h (hunk 1 above)
Comment (FonsRademakers) [inline core/unix/inc/TUnixSystem.h]: And please use ROOT coding standards.

Comment (bbockelm) [inline core/unix/inc/TUnixSystem.h]: (by the way, is there a good online ref for these to link to?)

**Comment by bbockelm** [comment]
Hi Zhe,

You may also want to pull in the work done here:

https://github.com/cms-sw/cmssw/pull/11662

Specifically,
a) Allow the read function to timeout.
b) Avoid use of execv on Linux.  CERN uses LD_PRELOAD tricks to insert a custom version of this function into the executable.  Unlike the POSIX documentation, the version of execv at CERN is not async signal safe.

Brian

File: core/unix/inc/TUnixSystem.h (hunk 1 above)
Comment (pcanal) [inline core/unix/inc/TUnixSystem.h]: Yes, of course :), at https://root.cern.ch/coding-conventions

Comment (bbockelm) [inline core/unix/inc/TUnixSystem.h]: Thanks Philippe - Zhe, can you run through this guide and try to get the code to compliance?

@pcanal - when contributing to CVMFS, they provide an automated tool (based on cpplint).  I see you reference astyle for managing whitespace.  Do you have anything similarly automated you use?

**Comment by zzxuanyuan** [comment]
@bbockelm Sure. I will polish up my code.

**Comment by zzxuanyuan** [comment]
@pcanal Could you take a look at the new patch? Is there any test cases I can run with?

Thanks!

File: core/unix/src/TUnixSystem.cxx [hunk 10]
```diff
@@ -568,6 +573,94 @@ TUnixSystem::~TUnixSystem()
 
 ////////////////////////////////////////////////////////////////////////////////
 /// Initialize Unix system interface.
+extern "C"
+{
+   static int FullWrite(int fd, const char *text)
+   {
+      const char *buffer = text;
+      size_t count = strlen(text);
+      ssize_t written = 0;
```
Comment (pcanal) [inline core/unix/src/TUnixSystem.cxx]: since written is not used outside the loop, it ought to be declared inside.

File: core/unix/src/TUnixSystem.cxx [hunk 11]
```diff
@@ -568,6 +573,94 @@ TUnixSystem::~TUnixSystem()
...
+      return 0;
+   }
+
+   static int FullErrWrite(const char *text) {
+      return FullWrite(2, text);
+   }
+
+}
+
+std::unique_ptr<std::thread> TUnixSystem::fHelperThread;
```
Comment (pcanal) [inline core/unix/src/TUnixSystem.cxx]: Static declarations ought to be towards the beginning of the files (before any function implementation).

File: core/unix/inc/TUnixSystem.h [hunk 12]
```diff
@@ -71,6 +78,17 @@ class TUnixSystem : public TSystem {
...
    static int          UnixSend(int sock, const void *buf, int len, int flag);
 
+   // added helper static members for stacktrace
+   static const int                    fPidStringLength = 255;
+   static char                         fPidString[fPidStringLength];
+   static int                          fParentToChild[2];
+   static int                          fChildToParent[2];
+   static std::unique_ptr<std::thread> fHelperThread;   
+   static char *const                  kStackArgv[];
+   static char *const *                GetStackArgv();
```
Comment (pcanal) [inline core/unix/inc/TUnixSystem.h]: This is a set of tightly related static, consider wrapping them into a private struct.

File: core/unix/src/TUnixSystem.cxx [hunk 13]
```diff
@@ -568,6 +573,94 @@ TUnixSystem::~TUnixSystem()
 
 ////////////////////////////////////////////////////////////////////////////////
 /// Initialize Unix system interface.
+extern "C"
+{
+   static int FullWrite(int fd, const char *text)
```
Comment (pcanal) [inline core/unix/src/TUnixSystem.cxx]: Why are those extern "C"?

Comment (pcanal) [inline core/unix/src/TUnixSystem.cxx]: FullWrite -> Could you tweak the name to be more specific of why it is different from other writing routines?

**Comment by pcanal** [comment]
> Is there any test cases I can run with?

Not specifically.  Some of the code would be exercised by calling ::Fatal( ... )

File: core/unix/src/TUnixSystem.cxx (hunk 13 above)
Comment (zzxuanyuan) [inline core/unix/src/TUnixSystem.cxx]: I referred to Brian's code in cmssw. I guess it is called "FullWrite" is because it has been used in signal handler and we want to make sure it is async-signal-safe (not interrupted in the middle).

Comment (pcanal) [inline core/unix/src/TUnixSystem.cxx]: Then I propose to call is SignalSafeWrite :)

thanks,
Philippe

Comment (zzxuanyuan) [inline core/unix/src/TUnixSystem.cxx]: Sure. It makes more sense. I will change it.

File: core/unix/inc/TUnixSystem.h (hunk 12 above)
Comment (zzxuanyuan) [inline core/unix/inc/TUnixSystem.h]: All the changes are done except this one. I might miss some coding standard of ROOT, is there any benefit to wrap them into a private struct?

Comment (pcanal) [inline core/unix/inc/TUnixSystem.h]: This is not part of standard.  The advantage would to more clearly segregate the functionality (even some of the function could be assign to that struct rather than TSystem itself).  This would actually make re-use of the code in a different context (a new version of TSystem for example) easier.

**Comment by pcanal** [comment]
Hi,

What was the result of your testing?  Did you create a test we can add to roottest?
Did  you already update to follow the coding conventions? (I do not see an newer version so I commented on what I see)

Thanks,
Philippe.

File: core/unix/inc/TUnixSystem.h [hunk 14]
```diff
@@ -31,11 +31,30 @@
...
 #endif
 
+#include <thread>
+#include <memory>
+
+int StackTraceExec(void *);
+
 typedef void (*SigHandler_t)(ESignals);
 
+struct TStackTraceHelper {
```
Comment (pcanal) [inline core/unix/inc/TUnixSystem.h]: Class not part of the public interface need to either be defined in an unnamed namepace in the source file or within the ROOT::Internal namespace or as protected/private subclass of the using class.

File: core/unix/inc/TUnixSystem.h [hunk 15]
```diff
@@ -31,11 +31,30 @@
...
+   static const int             stringLength = 255;
+   char                         shellExec[stringLength];
```
Comment (pcanal) [inline core/unix/inc/TUnixSystem.h]: data member must start with a lower f and be camel case words, for example fStringLength and fShellExec.

File: core/unix/inc/TUnixSystem.h [hunk 16]
```diff
@@ -71,6 +90,12 @@ class TUnixSystem : public TSystem {
    static int          UnixRecv(int sock, void *buf, int len, int flag);
    static int          UnixSend(int sock, const void *buf, int len, int flag);
 
+   // added helper static members for stacktrace
+   static char *const                  kStackArgv[];
```
Comment (pcanal) [inline core/unix/inc/TUnixSystem.h]: reduce the spacing between const and kStack to the minimal number to align the group of variable/functions names.

File: core/unix/src/TUnixSystem.cxx [hunk 17]
```diff
@@ -40,10 +40,15 @@
...
 
+#ifdef __linux__
+#include <syscall.h>
+#endif
+
 //#define G__OLDEXPAND
 
 #include <unistd.h>
 #include <stdlib.h>
+#include <poll.h>
```
Comment (pcanal) [inline core/unix/src/TUnixSystem.cxx]: I think this is not available on windows. Would using select be an option?  Otherwise we need to hide this (and code using it) behind 
```ifndef R__WIN32```

File: core/unix/src/TUnixSystem.cxx [hunk 18]
```diff
@@ -5148,3 +5275,133 @@ int TUnixSystem::GetProcInfo(ProcInfo_t *info) const
...
+       SignalSafeErrWrite(strerror(-result));
+       SignalSafeErrWrite("\n");
+       return;
+    }
+}
+
+void StackTraceFork()
+{
+   char childStack[4*1024];
+   char *childStackPtr = childStack + 4*1024;
```
Comment (pcanal) [inline core/unix/src/TUnixSystem.cxx]: if a constant is repeated, it ought to be declare as a constexpr to be used consistently.

File: core/unix/src/TUnixSystem.cxx [hunk 19]
```diff
@@ -387,6 +392,83 @@ class TFdSet {
...
+         if (errno == EINTR) { continue; }
+         else { return -errno; }
+      }
+      count -= written;
+      buffer += written;
+   }
+   return 0;
+}
+
+static int SignalSafeRead(int fd, char *inbuf, size_t len, int timeout=-1) {
```
Comment (pcanal) [inline core/unix/src/TUnixSystem.cxx]: missing function documentation.

File: core/unix/inc/TUnixSystem.h [hunk 20]
```diff
@@ -31,11 +31,30 @@
...
+
 typedef void (*SigHandler_t)(ESignals);
 
+struct TStackTraceHelper {
+   static const int             stringLength = 255;
+   char                         shellExec[stringLength];
+   char                         pidString[stringLength];
+   char                         pidNum[stringLength];
+   int                          parentToChild[2];
+   int                          childToParent[2];
```
Comment (pcanal) [inline core/unix/inc/TUnixSystem.h]: int should be referred to as 'Int_t'
char should be referred to as 'Char_t' 
(thought out the code).
the exception is 'const char*' (correctly spelled as 'const char*')

**Comment by pcanal** [comment]
Since all the signal handling and processing is essentially standalone, could be move it all in a single (sub)class?

Thanks,
Philippe.

File: core/unix/src/TUnixSystem.cxx (hunk 17 above)
Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: `select` could work, but it may be better to just disable the timeouts on Windows; on RHEL7 with some common GCC options, select triggers an assert when the FD is >= 1024.

(Is there a signal-safe read function elsewhere in ROOT that has timeout?  Seems silly that it is only used here.)

**Comment by bbockelm** [comment]
Hi Philippe,

Talked with Zhe about this yesterday; now that the end-of-semester rush is over, he's coming back to this item.  We're working to refactor this to a new class.

Brian

File: core/unix/src/TUnixSystem.cxx (hunk 17 above)
Comment (pcanal) [inline core/unix/src/TUnixSystem.cxx]: TSystem::Select uses ::select on linux and windows.

Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: Ah - TSystem::Select calculates its own bitmask instead of using `FD_SET` (see http://sourceforge.net/p/opalvoip/bugs/496/ for a decent explanation).  `FD_SET` is what causes the assert for large FDs.

Unfortunately, we can't use TSystem::Select because it internally uses `TFdSet`, which can lead to memory allocations (unsafe in the signal handler).

I guess we just have to use select on Windows and fail if fd > 1024.

**Comment by zzxuanyuan** [comment]
@pcanal @bbockelm I think there are two possible refactoring approach.
1. I could extract all signal handling related functions and attributes out from TSystem(.h,.cxx) and modify both TUnixSystem and TWinNT
2. I create SignalHandling class in another class and add it as an class member in TUnixSystem. In that sense, I need to overwrite all signal handling functions and point them to the corresponding functions in new class SignalHandling.

Zhe

**Comment by pcanal** [comment]
I meant a variation of 1, where TSystem calls directly the extracted methods (i.e. essentially this should just be 'refactoring').  Thanks.

**Comment by zzxuanyuan** [comment]
Sure. It makes more sense to me. The only concern I have is how much side effect it will introduce to windows. I just ignore it from now?

**Comment by pcanal** [comment]
well, it would be no more or less than what you have now :).   We will need to see how it behaves on Windows once we have merged it.

File: core/unix/src/TUnixSystem.cxx (hunk 17 above)
Comment (zzxuanyuan) [inline core/unix/src/TUnixSystem.cxx]: @bbockelm Since fSignals in TSystem has been also defined as TFdSet, should we change the data structure in the new signal handling?

Comment (bbockelm) [inline core/unix/src/TUnixSystem.cxx]: Nope, `fSignals` doesn't need to be changed because it isn't invoked from a signal handler.

Unix signal handlers have an extremely limited set of operations they are allowed to do: this is why we must be very careful to not call most ROOT functions.

**Comment by bbockelm** [comment]
@zzxuanyuan - since this appears to be replaced by #134, can you please close this request?
